win32ss/user/user32/windows/cursoricon.c

```diff
@@ -2041,16 +2041,53 @@ HANDLE WINAPI CopyImage(
     switch(uType)
     {
         case IMAGE_BITMAP:
+            if (!hImage)
+            {
+                SetLastError(ERROR_INVALID_HANDLE);
+                break;
+            }
             return BITMAP_CopyImage(hImage, cxDesired, cyDesired, fuFlags);
         case IMAGE_CURSOR:
         case IMAGE_ICON:
-        /* HACK: Copying bitmaps with LR_COPYFROMRESOURCE flag fails. CORE-17902.
-         * This is a way to return the original bit map if we need
-         * the icons to show up. We need a simpler test. */
         {
-            HANDLE handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
-            if (!handle && (fuFlags & (LR_COPYFROMRESOURCE|LR_COPYRETURNORG)))
-                handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, (fuFlags & ~LR_COPYFROMRESOURCE));
+            HANDLE handle;
+            if (!hImage)
+            {
+                SetLastError(ERROR_INVALID_CURSOR_HANDLE);
+                break;
+            }
+            handle = CURSORICON_CopyImage(hImage, uType == IMAGE_ICON, cxDesired, cyDesired, fuFlags);
+            if (!handle && (fuFlags & LR_COPYFROMRESOURCE))
+            {
+                /* Test if the hImage is the same size as what we want by getting
+                 * its BITMAPINFO and comparing it to the desired size. */
+                HDC hDC;
+                ICONINFO iconinfo = { 0 };
+                /* struct is for BITMAPINFO plus a full-sized color table */
+                struct
+                {
+                    BITMAPINFOHEADER bmiHeader;
+                    RGBQUAD bmiColors[256];
+                } bmi;
+
+                if (!GetIconInfo((HICON)hImage, &iconinfo))
+                    ERR("GetIconInfo Failed. hImage %p\n", hImage);
+                ZeroMemory(&bmi, sizeof(bmi));
+                bmi.bmiHeader.biSize = sizeof(bmi.bmiHeader);
+
+                hDC = GetDC(NULL);  // Screen hdc
```

Comment:
In principle this call shouldn't fail, but....

Reply:
@HBelusca, I did a non-exhaustive search of the ReactOS code base and in most cases where we use this code:
```
GetDC(NULL);
```
there is no error checking, but since you pointed this out, now I check if ```hDC``` is NULL.
If so, we will output an error message and skip the GetDIBits call that used this value.
We needed to free the ```iconinfo``` items, so we could not return here, but eventually after the failing "IF" test below, we will return "NULL".

@whindsaks, Sorry I do not understand your question about DIB_RGB_COLORS. Can you be more specific. Thanks.

Reply:
My question was, does GetDIBits need a hdc with DIB_RGB_COLORS or is it only really needed for DIB_PAL_COLORS (palette).

Reply:
@whindsaks, I do not really know the answer to your question, but it was a good one. In researching it, I found a much simpler way to get the width and height from an HBITMAP. So instead of using GetDIBits, I now use GetObject. So there is no longer any  use of the DIB_PAL_COLORS of DIB_RGB_COLORS involved here. Thanks.